### PR TITLE
Add ability to set frequency of emails

### LINKED ISSUE #5: allow multiple reminders
It would be nice if you could send out multiple reminders using the same activity - at the moment to do this requires multiple re-engagement activities to be set up.

### PR DESCRIPTION
**Pull request by danmarsden**
Fixes #5

### CODE REVIEW

**Comment by danmarsden** [comment]
@peterbulmer any chance you could review this patch? - adds support to send multiple e-mails with a few limits that seemed sensible.

File: db/install.xml [hunk 1]
```diff
@@ -20,6 +20,7 @@
         <FIELD NAME="emailsubjectmanager" TYPE="char" LENGTH="255" NOTNULL="true" SEQUENCE="false" COMMENT="Email subject line for emails going to user's manager"/>
         <FIELD NAME="emailrecipient" TYPE="int" LENGTH="4" NOTNULL="true" DEFAULT="0" SEQUENCE="false" COMMENT="Who should receive any email - user|manager|both"/>
         <FIELD NAME="duration" TYPE="int" LENGTH="10" NOTNULL="true" DEFAULT="604800" SEQUENCE="false" COMMENT="how long the reengagement is"/>
+        <FIELD NAME="emailfrequency" TYPE="int" LENGTH="3" NOTNULL="true" DEFAULT="1" SEQUENCE="false" COMMENT="how many times to send the email"/>
```
Comment (peterbulmer) [inline db/install.xml]: "frequency" to me means 'occurrences per second', (or per minute, hour etc).

I'd really like this to be called 'remindercount' or 'nagmagnitude' or something

File: mod_form.php [hunk 2]
```diff
@@ -292,4 +300,24 @@ public function add_completion_rules() {
     public function completion_rule_enabled($data) {
         return true;
     }
+    public function validation($data, $files) {
+        $errors = parent::validation($data, $files);
+
+        if (!empty($data['emailperiod']) && $data['emailperiod'] <= 60 && $data['emailfrequency'] > 5) {
+            // Email is sent based on minutes selector - make sure no more than 5 emails can be sent.
+            $errors['emailfrequency'] = get_string('frequencytoohigh', 'reengagement', 5);
```
Comment (peterbulmer) [inline mod_form.php]: I like your comprimse here re spam restrictions - avoiding the whole opt-out complexity & leaving it up to the course creator to sort out.

But I'd argue for a more restrictive threshold.  Allowing someone to spam 5 times over 25 minutes, or 12 times over a day seems a bit much.

File: mod_form.php [hunk 3]
```diff
@@ -292,4 +300,24 @@ public function add_completion_rules() {
...
+        } else if (!empty($data['emailperiod']) && $data['emailperiod'] <= 3600 && $data['emailfrequency'] > 12) {
+            // Email is sent based on hours selector - makes sure no more than 12 emails can be sent.
+            $errors['emailfrequency'] = get_string('frequencytoohigh', 'reengagement', 12);
+        } else if (!empty($data['emailperiod']) && $data['emailperiod'] <= 86400 && $data['emailfrequency'] > 24) {
+            // Email is sent based on days selctor - make sure no more than  24 emails can be sent.
+            $errors['emailfrequency'] = get_string('frequencytoohigh', 'reengagement', 24);
+        }
+
+        if ($data['emailperiod'] == 60 && $data['emailperiodcount'] < 5) {
+            $errors['emaildelay'] = get_string('periodtoolow', 'reengagement');
```
Comment (peterbulmer) [inline mod_form.php]: I recall that elsewhere a minimum email delay of 5 minutes is enforced.  If you set "1 minute" & click save, it saves 300 (5 minutes).

**Comment by peterbulmer** [comment]
I've put some inline comments in. But I also wanted an overall comment: Your patch assumes that the resend period is the same as the initial emaildelay.
I see a common use case that somebody would have 30 days to complete a quiz/scorm/whatever, and then get their first reminder.  Then if they don't complete it, they should be getting daily, or weekly reminders.
It's more involved, and requires more info to be stored 'email last sent' stored per user

File: mod_form.php (hunk 3 above)
Comment (danmarsden) [inline mod_form.php]: yes - but UI is not great - it just makes that change "silently" - it should really notify the user during validation right?

File: mod_form.php (hunk 2 above)
Comment (danmarsden) [inline mod_form.php]: I was thinking about the use-case of something "time-sensitive" - eg - a quiz that they must complete by a certain time. - but I think you're right - we should make them more restrictive.

File: db/install.xml (hunk 1 above)
Comment (danmarsden) [inline db/install.xml]: cool - I'll make that change - thanks.

**Comment by danmarsden** [comment]
thanks - I was thinking about simplicity of the patch - for your example they could set up 2 reengagments - one that does the initial delay, and then another that does the reminders?

**Comment by danmarsden** [comment]
rebased with those comments in mind - a future version could look at your extra use-case (or people could just use 2 reengagment activities to allow that scenario.) - let me know if you think this is ok to merge?

File: lib.php [hunk 4]
```diff
@@ -322,7 +322,7 @@ function reengagement_cron() {
                   INNER JOIN {reengagement} r on r.id = ip.reengagement
                        WHERE ip.emailtime < :emailtime
                              AND r.emailuser = " . REENGAGEMENT_EMAILUSER_TIME . '
-                             AND ip.emailsent = 0
+                             AND ip.emailsent <= r.remindercount
```
Comment (peterbulmer) [inline lib.php]: I think you're out-by-one here:
if remindercount is 2,
and a user has been sent 2 emails, that record will still be caught.

**Comment by peterbulmer** [comment]
Otherwise, looks good.

Thanks!

**Comment by danmarsden** [comment]
thanks for the review! - good spotting, it would have returned all records instead of just the ones that need to be dealt with. I've merged that in now.

**Pull request merged**